libs/net/kafka/KafkaListener`.cs

```diff
@@ -193,20 +216,57 @@ public async Task ConsumeAsync(Func<ConsumeResult<TKey, TValue>, Task<ConsumerAc
         }
         catch (Exception ex)
         {
-            // https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#fatal-consumer-errors
-            // An unhandled exception will stop consuming as I don't know of a way to resume the paused consumer.
-            _logger.LogError(ex, "Error while consuming. {message}", ex.Message);
-            proceed = this.OnError?.Invoke(this, new ErrorEventArgs(ex)) ?? ConsumerAction.Stop;
+            proceed = HandleException(ex);
         }
         finally
         {
-            if (proceed == ConsumerAction.Stop) this.Stop();
-            else if (this.IsPaused && proceed == ConsumerAction.Proceed)
+            if (!UseMultiThreads) FinalCleanUp(proceed);
+        }
+    }
+
+    private async Task<ConsumerAction> RunActionAsync(
+        Func<ConsumeResult<TKey, TValue>, Task<ConsumerAction>> action,
+        ConsumeResult<TKey, TValue>? _currentResult)
+    {
+        var result = await action(_currentResult!);
+        if (result == ConsumerAction.Proceed)
+        {
+            // Manually commit to inform Kafka this message has successfully been handled.
+            // TODO: Convert to more efficient process and use EnableAutoOffsetStore=false - https://docs.confluent.io/kafka-clients/dotnet/current/overview.html#store-offsets
+            if (_config.EnableAutoCommit == false)
             {
-                _logger.LogDebug("Resuming consumption: {topics}", String.Join(", ", this.Consumer!.Subscription ?? new List<string>()));
-                this.IsPaused = false;
-                this.Consumer?.Resume(this.Consumer.Assignment);
+                this.Consumer?.Commit(_currentResult);
+                _logger.LogDebug("Message committed from Kafka topic:'{topic}' key:'{key}'", _currentResult!.Topic, _currentResult.Message.Key);
             }
+            
+            _currentResult = null;
+        }
+        return result;
+    }
+
+    private ConsumerAction HandleException(Exception ex)
+    {
+        // https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#fatal-consumer-errors
+        // An unhandled exception will stop consuming as I don't know of a way to resume the paused consumer.
+        _logger.LogError(ex, "Error while consuming. {message}", ex.Message);
+        return this.OnError?.Invoke(this, new ErrorEventArgs(ex)) ?? ConsumerAction.Stop;
+    }
+
+    private void FinalCleanUp(ConsumerAction proceed, CancellationToken cancellationToken = default)
+    {
+        if (proceed == ConsumerAction.Stop)
+        {
+            if (UseMultiThreads && !_currentResults.IsEmpty)
+            {
+                _currentResults.Clear();
```

Comment:
If you clear all the results as the result of one action failing and asking for a stop it will result in unexpected behaviour in all the threads that are running.